sysprog21/semu
# Implement VirtIO sound device playback

Implement VirtIO sound device supporting these operations (the item with checked box checked means it is implemented right now):

- For setting up the device:
  - [x] `VIRTIO_SND_R_JACK_INFO`
  - [x] `VIRTIO_SND_R_PCM_INFO`
  - [x] `VIRTIO_SND_R_CHMAP_INFO`
- For playing the sound (PCM):
  - [x] `VIRTIO_SND_R_PCM_SET_PARAMS`
  - [x] `VIRTIO_SND_R_PCM_PREPARE`
  - [x] `VIRTIO_SND_R_PCM_RELEASE`
  - [x] `VIRTIO_SND_R_PCM_START`
  - [x] `VIRTIO_SND_R_PCM_STOP`

## Limitations

1. The emulator will hang if PulseAudio is enabled on host.
    It causes to CNFA and CNFA cannot initialize if PulseAudio
    is disabled, What's worse, it needs re-working as CNFA uses
    different threading models between pure ALSA and PulseAudio environment.
    Therefore, users need mitigations (i.e., restart the emulator after
    playing sound) or just wait for the future release.
    
2. The playback may plays with repeating artifact (for example,
    A "front center" ALSA example sound will sound like "front front
    center"). The root cause is the Linux Kernel and it hasn't got
    fixed even in mainline version.

## Test Cases

### boot up test

#### test procedures

1. Execute `make check` to run `semu`.
2. Check kernel message (`dmesg`).

#### expected results

1. The following message should appear while booting up:

```
[    4.011962] ALSA device list:
[    4.015962]   #0: Loopback 1
[    4.015962]   #1: VirtIO SoundCard at platform/f4400000.virtio/virtio2
```

### check driver configuration

#### test procedures

1. Execute `aplay -l` in emulator.
2. Check the output messages.

#### expected results

1. The following message should appear after Step 1:

```
$ aplay -l          
**** List of PLAYBACK Hardware Devices ****

<other sound device here>

card 1: SoundCard [VirtIO SoundCard], device 0: virtio-snd [VirtIO PCM 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```

### play sound

#### test procedures

1. Execute `speaker-test` in emulator.
4. Check the host speaker.

#### expected results

1. A white noise will be played by host speaker while `speaker-test` is executing.

### play sequence sound

#### test procedures

1. Execute `aplay /usr/share/sounds/alsa/Front_Center.wav` in emulator.
2. Check the host speaker.

#### expected results

1. A sound with "Front Center" speaking should be played.
